File: backend/src/main/java/com/example/contact/domain/valueobject/Email.java
package com.example.contact.domain.valueobject;

import com.example.contact.domain.exception.InvalidContactDataException;
import java.util.Locale;
import java.util.regex.Pattern;

/**
 * Email address value object with normalization to lowercase trimmed form.
 */
public final class Email {

    private static final Pattern FORMAT =
            Pattern.compile("^[\\w.+-]+@[\\w.-]+\\.[A-Za-z]{2,}$");

    private final String value;

    private Email(String value) {
        this.value = value;
    }

    /**
     * Creates an email from raw input after trim and lowercase normalization.
     *
     * @param raw user input
     * @return normalized email
     * @throws InvalidContactDataException when blank or malformed
     */
    public static Email create(String raw) {
        if (raw == null || raw.isBlank()) {
            throw new InvalidContactDataException("Email must not be blank");
        }
        String normalized = raw.trim().toLowerCase(Locale.ROOT);
        if (!FORMAT.matcher(normalized).matches()) {
            throw new InvalidContactDataException("Invalid email format");
        }
        return new Email(normalized);
    }

    /**
     * Rehydrates an email from persistence without re-validating format.
     *
     * @param stored stored value
     * @return email instance
     */
    public static Email fromStored(String stored) {
        return new Email(stored);
    }

    public String value() {
        return value;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Email email)) {
            return false;
        }
        return value.equals(email.value);
    }

    @Override
    public int hashCode() {
        return value.hashCode();
    }

    @Override
    public String toString() {
        return value;
    }
}

File: frontend/src/contact/domain/valueobject/Email.ts
import { InvalidContactDataError } from '../exception/InvalidContactDataError';

const EMAIL_PATTERN = /^[\w.+-]+@[\w.-]+\.[A-Za-z]{2,}$/;

/**
 * Email value object with normalization to lowercase trimmed form.
 */
export class Email {
  private constructor(private readonly value: string) {}

  /**
   * Creates an email from raw input after validation and normalization.
   * @param raw - user input
   * @returns normalized email
   * @throws InvalidContactDataError when blank or malformed
   */
  static create(raw: string): Email {
    if (!raw?.trim()) {
      throw new InvalidContactDataError('Email must not be blank');
    }
    const normalized = raw.trim().toLowerCase();
    if (!EMAIL_PATTERN.test(normalized)) {
      throw new InvalidContactDataError('Invalid email format');
    }
    return new Email(normalized);
  }

  /**
   * Rehydrates from stored value without re-validation.
   * @param stored - persisted email
   * @returns email instance
   */
  static fromStored(stored: string): Email {
    return new Email(stored);
  }

  /**
   * Returns the email string value.
   * @returns email address
   */
  toString(): string {
    return this.value;
  }
}
